### PR TITLE
shrink{h,v}: improve performance

### PR DESCRIPTION
**Pull request by kleisauke**
- Commit 4395dd0a098ac626bef5b8f1673272e18fa3010c is a revert of commit 9b968491d366a68ea1008ed29a74f7293501e4c4. See https://gist.github.com/kleisauke/dcf24beb95cdf2b3037ff63f71b7f9bd for the rationale behind this.
- Commit e1698911402a2f3caac7131e37a8e438e1ab0827 takes a similiar approach as PR #3052 in `shrinkh`.
- Commit 77e8451f13c0aeb949648b299a4148eaec2584a9 removes the `shrinkv` sum buffer in favor of a local, ~7% faster. In line with commit 2e5880b69cf79715ac84abb28b8482aefe7df532.
- Commit 3d28a9222fcf8cd4b49f5b20e716096e60c22b38 uses `gint64` as accumulator type for the uint/int paths. In line with PR #3514.
- Commit 2d8f1c6a1b31e3cb6aef2282f83ce0f66d66778a adds a fixed-point arithmetic path for uchar images, ~9% faster.

<details>
  <summary>Test script</summary>

Test image: https://github.com/kleisauke/vips-microbench/blob/master/images/x.jpg
```python
#!/usr/bin/env python

import sys
import pyvips

im = pyvips.Image.new_from_file(sys.argv[1])

for fac in [2, 4, 8, 16]:
    for fmt in ["uchar", "ushort", "uint",
                "char", "short", "int",
                "float", "double",
                "complex", "dpcomplex"]:
        x = im.cast(fmt, shift=True)
        s = x.shrink(fac, fac)
        d = abs(s.avg() - x.avg())
        print(d, fmt)
        assert d < 2
```
</details>

### CODE REVIEW

File: libvips/resample/shrinkh.c [hunk 1]
```diff
@@ -85,13 +81,10 @@ G_DEFINE_TYPE(VipsShrinkh, vips_shrinkh, VIPS_TYPE_RESAMPLE);
...
-\
-				sum = 0; \
-				x1 = b; \
-				VIPS_UNROLL(shrink->hshrink, INNER(BANDS)); \
-				q[b] = (sum + shrink->hshrink / 2) / \
-					shrink->hshrink; \
+				ACC_TYPE sum = amend; \
+				for (x1 = b; x1 < ne; x1 += BANDS) \
+					sum += p[x1]; \
+				q[b] = (sum * multiplier) >> bitshift; \
```
[the remark this answers is not in the document]
Comment (kleisauke) [reply libvips/resample/shrinkh.c]: You're right, this bitwise trickery is likely prone to overflow, underflow, and rounding errors, especially for non-uchar images.

For reference, this was inspired by Pillow:
https://github.com/python-pillow/Pillow/blob/0995305eb4f991c71e157a04c731af7c7ee9785c/src/libImaging/Reduce.c#L48

[the remark this answers is not in the document]
Comment (jcupitt) [reply libvips/resample/shrinkh.c]: Ah OK. Yes, libvips has fixed-point arithmetic in a few places too heh

**Comment by jcupitt** [comment]
You could keep the mul + shift for a fast uint8 path I guess. You could do uint16 and uint32 as well, but that's such a rare case it's probably not worth it.

This is a very unintuitive speedup for me! I was sure the accumulator array would be faster, since it would vectorise so well and would fetch pixel values in strict memory order. There must be something important I don't understand about cache lines :(

**Comment by jcupitt** [comment]
I built master with:

```
meson setup build --prefix ~/vips --libdir lib --optimization 3
```

ie. enable the autovectoriser, then with a 10k x 10k pixel RGB JPG I ran some tests like this:

```
$ time vips shrinkv wtc.jpg x.v 128

real	0m0.400s
user	0m0.326s
sys	0m0.082s
```

then fastest of five runs.

branch | vshrink | user time (s)
------ | ------- | -------------
master | 2       | 1.01
master | 10      | 0.45
master | 100     | 0.33
this   | 2       | 0.94
this   | 10      | 0.43
this   | 100     | 0.43

So I suppose I'd interpret that as being: 

- somewhere under 100 scanlines, the whole block being shrunk fits in L1 cache and the non-local fetches that the vertical accumulator loop performs fit, meaning non-locality doesn't matter

- at 100 and over, we're seeing some cache thrashing and the new vertical loop is a little slower

- the autovectoriser on the horizontal loop along the scanline seems to have no effect on runtime, which is strange (to me)

**Comment by kleisauke** [comment]
Hmm, ASan seems to be quite happy without that `vips_embed()` change. I wonder why it complained on this changeset though:
https://github.com/kleisauke/libvips/compare/optimize-shrink...shrink-highway
https://github.com/kleisauke/libvips/actions/runs/12050557645/job/33599676999

Anyways, I think the updated revision of this PR should fix the performance degradation for large `vshrink`'s.

**Comment by kmartinez** [comment]
Oh wow have we come full circle? John and I spent ages optimising older vips code to avoid multiply and especially divide (as they were so slow!) using shifts, look-up tables etc. With today's fast cores its easy to forget it still takes > 10 clocks to do a divide ;-)

**Comment by jcupitt** [comment]
I read it again, let's do this! Nice job Kleis.

**Pull request merged**